README.md
# SpotBugs Gradle Plugin

[![Build Status](https://travis-ci.org/spotbugs/spotbugs-gradle-plugin.svg?branch=master)](https://travis-ci.org/spotbugs/spotbugs-gradle-plugin)
[![Coverage Status](https://sonarcloud.io/api/project_badges/measure?project=com.github.spotbugs.gradle&metric=coverage)](https://sonarcloud.io/component_measures?id=com.github.spotbugs.gradle&metric=coverage)
[![Debt](https://sonarcloud.io/api/project_badges/measure?project=com.github.spotbugs.gradle&metric=sqale_index)](https://sonarcloud.io/component_measures/domain/Maintainability?id=com.github.spotbugs.gradle)
[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

## License

Copyright 2017-2019 the original author or authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md

# Changelog

This is the changelog for SpotBugs Gradle Plugin. This follows [Keep a Changelog v1.0.0](http://keepachangelog.com/en/1.0.0/).

Currently the versioning policy of this project follows [Semantic Versioning](http://semver.org/) from version 1.6.0.

## Unreleased - 2019-??-??

### Fixed

* Deprecation warning in Gradle 6 ([#171](https://github.com/spotbugs/spotbugs-gradle-plugin/issues/171))

## 3.0.0 - 2019-11-14

### Added

* Gradle 6.1 support ([#174](https://github.com/spotbugs/spotbugs-gradle-plugin/issues/174))

### Removed

* Gradle 5.1 support ([#174](https://github.com/spotbugs/spotbugs-gradle-plugin/issues/174))

## 2.0.1 - 2019-10-16

### Fixed

* Avoid needless task creation (#165)

## 2.0.0 - 2019-05-13

### Removed

* Gradle 5.0 support

## 1.7.1 - 2019-03-14

### Fixed

* Treat `projectName` and `release` as task input ([#131](https://github.com/spotbugs/spotbugs-gradle-plugin/issues/131))

## 1.7.0 - 2019-03-14

### Added

* Add `-projectName` option ([#111](https://github.com/spotbugs/spotbugs-gradle-plugin/issues/111))
* Add `-release` option ([#111](https://github.com/spotbugs/spotbugs-gradle-plugin/issues/111))
* Add `setStylesheet(String)` that accepts relative path from the root of `spotbugs.jar` ([#107](https://github.com/spotbugs/spotbugs-gradle-plugin/issues/107))

## 1.6.11 - 2019-03-10

### Changed

* Use SpotBugs 3.1.12 by default
* Upgrade SLF4J to 1.8.0-beta4

## 1.6.10 - 2019-02-18

### Fixed

* "Trying to add already registered factory" problem reported as [a spotbugs issue](https://github.com/spotbugs/spotbugs/issues/819)
* Result caching works across checkouts. Previously it was using absolute paths and therefore didn't work. ([#96](https://github.com/spotbugs/spotbugs-gradle-plugin/pull/96))
* Support text type bug report ([#31](https://github.com/spotbugs/spotbugs-gradle-plugin/issues/31))

### Removed

* Usage of worker API introduced at [#57](https://github.com/spotbugs/spotbugs-gradle-plugin/issues/57)
* Drop support for Gradle 4, that causes [SLF4J related problem](https://github.com/gradle/gradle/issues/2657)

### Changed

* Use SpotBugs 3.1.11 by default

## 1.6.9 - 2019-01-04

### Fixed

* `classes` filtering that was broken at 1.6.2 ([#54](https://github.com/spotbugs/spotbugs-gradle-plugin/issues/54))

## 1.6.8 - 2018-12-21

### Changed

* Use SpotBugs 3.1.10 by default

## 1.6.7 - 2018-12-18

### Changed

* Keep track of top-level source set folders [#59](https://github.com/spotbugs/spotbugs-gradle-plugin/pull/59)

### Fixed

* Analyze runtime classpath instead of compile classpath. [#62](https://github.com/spotbugs/spotbugs-gradle-plugin/issues/62)

## 1.6.6 - 2018-12-08

## Fixed

* Prevent NPE in unexpected situations where no reports are enabled, despite xml being set by default using convention mapping (See [#61](https://github.com/spotbugs/spotbugs-gradle-plugin/issues/61), [#68](https://github.com/spotbugs/spotbugs-gradle-plugin/issues/68))

## 1.6.5 - 2018-10-17

### Changed

* Replace usage of internal Gradle APIs with supported [Worker API](https://guides.gradle.org/using-the-worker-api/). [#58](https://github.com/spotbugs/spotbugs-gradle-plugin/pull/58)

## 1.6.4 - 2018-09-26

### Changed

* Use SpotBugs 3.1.7 by default
* Replace usage of internal class ClosureBackedAction [#51](https://github.com/spotbugs/spotbugs-gradle-plugin/pull/51)

## 1.6.3 - 2018-09-08

* Use SpotBugs 3.1.6
* Added support for passing JVM arguments to the JavaExecHandleBuilder when creating a SpotBugsTask.
* Support Gradle 4.10 that cause problem with problematic method reference. ([#40](https://github.com/spotbugs/spotbugs-gradle-plugin/pull/40))

## 1.6.2 - 2018-05-23

* Use SpotBugs 3.1.3
* Add `showProgress` option to task mapping.
* Added a minor fix to handle multiple output source directories. With the release of gradle 4.0 the default location of classes changed from "build/classes/main" -> "build/{java, groovy, scala}/main". This caused a bug where the plugin would ignore any classes compiled which were not located within "build/classes/" or for clean groovy/scala projects the plugin would outright fail to perform an analysis due to NO-SOURCE.([#12](https://github.com/spotbugs/spotbugs-gradle-plugin/pull/12))
* Support Gradle 4.8 [#22](https://github.com/spotbugs/spotbugs-gradle-plugin/pull/22)

## 1.6.1 - 2018-02-25

* Use SpotBugs 3.1.2
* Fixed "loud output" [#506](https://github.com/spotbugs/spotbugs/issues/506) by adding an option to show progress and disabling it by default, matching the behaviour of the FindBugs Gradle Plugin

## 1.6.0 - 2017-10-25

* Use SpotBugs 3.1.0

## 1.5 - 2017-10-14

* Use SpotBugs 3.1.0-RC7
* Make build failed when user uses unsupported Gradle version ([#357](https://github.com/spotbugs/spotbugs/issues/357))
* Make error message human readable ([#428](https://github.com/spotbugs/spotbugs/pull/428))
* Fix missing dependency on compile task ([#440](https://github.com/spotbugs/spotbugs/issues/440))

## 1.4 - 2017-09-25

* Use SpotBugs 3.1.0-RC6
* Fix "Cannot convert the provided notation to a File or URI: classesDirs" error ([#320](https://github.com/spotbugs/spotbugs/issues/320))
* Support working with Android Gradle Plugin 2.3 ([#256](https://github.com/spotbugs/spotbugs/issues/256))

## 1.3 - 2017-08-16 [YANKED]

* Use SpotBugs 3.1.0-RC5
* Stop using [single class directory](https://docs.gradle.org/4.0.2/release-notes.html#multiple-class-directories-for-a-single-source-set) to prepare for Gradle v5 ([#299](https://github.com/spotbugs/spotbugs/issues/299))
* Print 'SpotBugs' instead of 'FindBugs' ([#291](https://github.com/spotbugs/spotbugs/issues/291))

## 1.2 - 2017-07-21

* Use SpotBugs 3.1.0-RC4
* Fixed [#214](https://github.com/spotbugs/spotbugs/issues/214)

## 1.1 - 2017-06-10

* Use SpotBugs 3.1.0-RC2

## 1.0 - 2017-05-16

* First release which uses FindBugs 3.0.1


src/main/java/com/github/spotbugs/SpotBugsPlugin.java
package com.github.spotbugs;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;
import java.util.Properties;
import java.util.concurrent.Callable;

import org.gradle.api.artifacts.Configuration;
import org.gradle.api.artifacts.DependencySet;
import org.gradle.api.file.FileCollection;
import org.gradle.api.internal.ConventionMapping;
import org.gradle.api.plugins.quality.CodeQualityExtension;
import org.gradle.api.plugins.quality.internal.AbstractCodeQualityPlugin;
import org.gradle.api.reporting.SingleFileReport;
import org.gradle.api.tasks.SourceSet;
import org.gradle.util.GradleVersion;

/**
 * A plugin for the <a href="https://spotbugs.github.io">SpotBugs</a> byte code analyzer.
 *
 * <p>
 * Declares a <tt>spotbugs</tt> configuration which needs to be configured with the SpotBugs library to be used.
 * Additional plugins can be added to the <tt>spotbugsPlugins</tt> configuration.
 *
 * <p>
 * For projects that have the Java (base) plugin applied, a {@link SpotBugsTask} task is
 * created for each source set.
 *
 * @see SpotBugsTask
 * @see SpotBugsExtension
 */
public class SpotBugsPlugin extends AbstractCodeQualityPlugin<SpotBugsTask> {

    /**
     * Supported Gradle version described at <a href="http://spotbugs.readthedocs.io/en/latest/gradle.html">official
     * manual site</a>.
     *
     * Package-protected access is for testing purposes
     */
    static final GradleVersion SUPPORTED_VERSION = GradleVersion.version("5.2");

    private SpotBugsExtension extension;

    @Override
    protected String getToolName() {
        return "SpotBugs";
    }

    @Override
    protected Class<SpotBugsTask> getTaskType() {
        return SpotBugsTask.class;
    }

    @Override
    protected void beforeApply() {
        verifyGradleVersion(GradleVersion.current());
        configureSpotBugsConfigurations();
    }

    /**
     * Verify that given version is supported by {@link SpotBugsPlugin} or not.
     *
     * @param version
     *            to verify
     * @throws IllegalArgumentException
     *             if given version is not supported
     */
    void verifyGradleVersion(GradleVersion version) throws IllegalArgumentException {
        if (version.compareTo(SUPPORTED_VERSION) < 0) {
            String message = String.format("Gradle version %s is unsupported. Please use %s or later.", version,
                    SUPPORTED_VERSION);
            throw new IllegalArgumentException(message);
        }
    }

    private void configureSpotBugsConfigurations() {
        Configuration configuration = project.getConfigurations().create("spotbugsPlugins");
        configuration.setVisible(false);
        configuration.setTransitive(true);
        configuration.setDescription("The SpotBugs plugins to be used for this project.");
    }

    @Override
    protected CodeQualityExtension createExtension() {
        extension = project.getExtensions().create("spotbugs", SpotBugsExtension.class, project);
        extension.setToolVersion(loadToolVersion());
        return extension;
    }

    String loadToolVersion() {
        return loadVersion("spotbugs-version");
    }

    String loadSlf4jVersion() {
        return loadVersion("slf4j-version");
    }

    private String loadVersion(String name) {
        URL url = SpotBugsPlugin.class.getClassLoader().getResource("spotbugs-gradle-plugin.properties");
        try (InputStream input = url.openStream()) {
            Properties prop = new Properties();
            prop.load(input);
            return prop.getProperty(name);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    @Override
    protected void configureTaskDefaults(SpotBugsTask task, String baseName) {
        task.setPluginClasspath(project.getConfigurations().getAt("spotbugsPlugins"));
        Configuration configuration = project.getConfigurations().getAt("spotbugs");
        configureDefaultDependencies(configuration);
        configureTaskConventionMapping(configuration, task);
        configureReportsConventionMapping(task, baseName);
    }

    protected void configureConfiguration(Configuration configuration) {
        // For an abstract method that was newly defined from v4.8, we need an empty method at here
        // https://github.com/spotbugs/spotbugs-gradle-plugin/issues/22
    }

    /**
     * Overriding this method, to include SLF4J into {@code spotbugsClasspath}. SLF4J is necessary in worker process.
     */
    @Override
    protected void createConfigurations() {
        Configuration configuration = project.getConfigurations().create(getConfigurationName());
        configuration.setVisible(false);
        configuration.setTransitive(true);
        configuration.setDescription("The " + getToolName() + " libraries to be used for this project.");
        // Don't need these things, they're provided by the runtime
        configuration.exclude(excludeProperties("ant", "ant"));
        configuration.exclude(excludeProperties("org.apache.ant", "ant"));
        configuration.exclude(excludeProperties("org.apache.ant", "ant-launcher"));
        configuration.exclude(excludeProperties("org.slf4j", "jcl-over-slf4j"));
        configuration.exclude(excludeProperties("org.slf4j", "log4j-over-slf4j"));
        configuration.exclude(excludeProperties("commons-logging", "commons-logging"));
        configuration.exclude(excludeProperties("log4j", "log4j"));
        configureConfiguration(configuration);
    }

    private Map<String, String> excludeProperties(String group, String module) {
        Map<String, String> map = new HashMap<>();
        map.put("group", group);
        map.put("module", module);
        return map;
    }
    private void configureDefaultDependencies(Configuration configuration) {
        configuration.defaultDependencies((DependencySet dependencies) -> {
            dependencies.add(project.getDependencies().create("org.slf4j:slf4j-simple:" + loadSlf4jVersion()));
            dependencies.add(project.getDependencies().create("com.github.spotbugs:spotbugs:" + extension.getToolVersion()));
        });
    }

    private void configureTaskConventionMapping(Configuration configuration, SpotBugsTask task) {
        ConventionMapping taskMapping = task.getConventionMapping();
        taskMapping.map("spotbugsClasspath", () -> configuration);
        taskMapping.map("ignoreFailures", extension::isIgnoreFailures);
        taskMapping.map("effort", extension::getEffort);
        taskMapping.map("reportLevel", extension::getReportLevel);
        taskMapping.map("visitors", extension::getVisitors);
        taskMapping.map("omitVisitors", extension::getOmitVisitors);

        taskMapping.map("excludeFilterConfig", extension::getExcludeFilterConfig);
        taskMapping.map("includeFilterConfig", extension::getIncludeFilterConfig);
        taskMapping.map("excludeBugsFilterConfig", extension::getExcludeBugsFilterConfig);

        taskMapping.map("extraArgs", extension::getExtraArgs);
        taskMapping.map("showProgress", extension::isShowProgress);

        taskMapping.map("jvmArgs", extension::getJvmArgs);
    }

    private void configureReportsConventionMapping(SpotBugsTask task, final String baseName) {
        task.getReports().all((final SingleFileReport report) -> {
            ConventionMapping reportMapping = conventionMappingOf(report);
            reportMapping.map("enabled", () -> report.getName().equals("xml"));
            reportMapping.map("destination", () -> new File(extension.getReportsDir(), baseName + "." + report.getName()));
        });
    }

    @Override
    protected void configureForSourceSet(final SourceSet sourceSet, SpotBugsTask task) {
        task.setDescription("Run SpotBugs analysis for " + sourceSet.getName() + " classes");
        task.setSourceSet(sourceSet);
        ConventionMapping taskMapping = task.getConventionMapping();
        taskMapping.map("classes", (Callable<FileCollection>) () -> {
            /*
             * As a result of the changes made in gradle 4.0.
             * See https://docs.gradle.org/4.0/release-notes.html - Location of classes in the build directory
             * Compile no longer bundles all classes in one directory build-gradle/classes/main
             * but instead separates classes into build-gradle/classes/{language}/main.
             *
             * We must therefor retrieve all output directories. Filter away the once that don't exist. Add each
             * existing file tree dependency to specified task. And then return the complete fileCollection, contain
             * all .class files available for analysis.
             */
            FileCollection presentClassDirs = sourceSet.getOutput().getClassesDirs().filter(File::exists);
            return presentClassDirs.getAsFileTree();
        });
        taskMapping.map("classpath", sourceSet::getRuntimeClasspath);
    }
}
